keras_fsl/dataframe/operators/to_k_shot_dataset.py

```diff
@@ -25,49 +37,78 @@ def __init__(self, k_shot, preprocessing, label_column="label_one_hot", cache=No
         self.preprocessing = preprocessing
         self.label_column = label_column
         self.cache = cache
+        self._reset_cache = reset_cache
+        self._max_shuffle_buffer_size = max_shuffle_buffer_size
+
+        _transforms = {
+            "with_tf_record": self.to_dataset_with_tf_record,
+            "with_cache": self.to_dataset_with_cache,
+            "raw": self.to_dataset_direct,
+        }
+        self._transform_group_to_dataset = _transforms[dataset_mode]
         if cache is not None:
             self.cache = Path(cache)
             self.cache.mkdir(exist_ok=True)
 
+    def repeat_k_shot(self, index):
+        return tf.data.Dataset.from_tensors(index).repeat(self.k_shot)
+
+    def transform_group_to_shuffled_dataset(self, group):
+        return (
+            self._transform_group_to_dataset(group)
+            .shuffle(buffer_size=min(self._max_shuffle_buffer_size, len(group)), reshuffle_each_iteration=True)
+            .repeat()
+        )
+
     @staticmethod
-    def load_img(annotation):
+    def to_dataset_direct(group):
         """
-        Args:
-            annotation (dict): with keys 'image_name': path to the image and 'crop_window' to be passed to tf.io.decode_and_crop_jpeg
-        Returns:
-            dict: the input dict with an extra image key.
+        Transform a pd.DataFrame into a tf.data.Dataset and load images
         """
-        return {
-            "image": tf.io.decode_and_crop_jpeg(
-                tf.io.read_file(annotation["image_name"]), crop_window=annotation["crop_window"], channels=3,
-            ),
-            **annotation,
-        }
+        return tf.data.Dataset.from_tensor_slices(group.to_dict("list")).map(
+            ip.add_field(ip.load_crop_as_ndarray, "image"), num_parallel_calls=tf.data.experimental.AUTOTUNE
+        )
 
-    def repeat_k_shot(self, index):
-        return tf.data.Dataset.from_tensors(index).repeat(self.k_shot)
+    def to_dataset_with_cache(self, group):
+        """
+        Transform a pd.DataFrame into a tf.data.Dataset and load images
+        """
+        return self.to_dataset_direct(group).cache(str(self.cache / group.name))
 
-    def to_dataset(self, group):
+    def to_dataset_with_tf_record(self, group):
         """
         Transform a pd.DataFrame into a tf.data.Dataset and load images
         """
-        return (
+        filename = self.cache / group.name
+        original_dataset = (
             tf.data.Dataset.from_tensor_slices(group.to_dict("list"))
```

Comment:
pourquoi on n'appelle pas `to_dataset_direct` ?

Reply:
Je pense que ça se relit à ton autre commentaire

> je n'aime pas trop faire des utils qui appellent des utils avec une étape de plus. On n'a pas un autre moyen de faire ça ?

Dans ma tête, il y a un overhead à appliquer 2 maps avec les autotunes.
Mais c'est probablement négligeable.

On peut réutiliser to_dataset_direct puis appeler une variation de encode_jpeg.

Quelque chose dans le style de :

```py
        original_dataset = (
            self.to_dataset_direct(group)
            .map(ip.add_field(ip.apply_on_field(tf.image.encode_jpeg, "image"), "image"), num_parallel_calls=tf.data.experimental.AUTOTUNE)
        )
```

Reply:
peut être un change_field

Reply:
Le truc que j'aimerai bien c'est qu'on puisse réutiliser ip pour d'autres façon de créer le dataset.

Reply:
est-ce que ça te semble mieux avec cette itération ?